src/TodosContext.tsx

```diff
@@ -0,0 +1,209 @@
+/* eslint-disable no-console */
+import React, { useEffect, useState } from 'react';
+import { Filter } from './services/enums';
+import {
+  addTodo,
+  deleteTodo,
+  getTodos,
+} from './api/todos';
+import { Todo } from './types';
+
+const USER_ID = 11138;
+
+interface FilterParams {
+  filterBy?: Filter,
+}
+
+function filterTodosByCompleted(todos: Todo[], filter: Filter): Todo[] {
+  let todosCopy = [...todos];
+
+  switch (filter) {
+    case (Filter.ACTIVE): {
+      todosCopy = todosCopy.filter(todo => !todo.completed);
+      break;
+    }
+
+    case (Filter.COMPLETED): {
+      todosCopy = todosCopy.filter(todo => todo.completed);
+      break;
+    }
+
+    default: {
+      break;
+    }
+  }
+
+  return todosCopy;
+}
+
+function filterTodos(todos: Todo[], { filterBy }: FilterParams): Todo[] {
+  let todosCopy = [...todos];
+
+  if (filterBy) {
+    todosCopy = filterTodosByCompleted(todos, filterBy);
+  }
+
+  return todosCopy;
+}
+
+function deleteAllTodos(todosIds: number[]): Promise<boolean> {
+  return new Promise((resolve) => {
+    todosIds.forEach(id => deleteTodo(id)
+      .then(() => resolve(true)));
+  });
+}
+
+interface ContextProps {
+  todos: Todo[],
+  visibleTodos: Todo[],
+  todoAdd: (newQuery: string) => Promise<boolean>,
+  clearAllCompleted: () => void,
+  todoDelete: (todoId: number) => Promise<boolean>,
+  isTodosHasCompleted: () => boolean,
+  isEveryTodoCompleted: () => boolean,
+  filterBy: Filter,
+  setFilterBy: (newFilter: Filter) => void,
+  errorMessage: string,
+  areCompletedDeletingNow: boolean,
+  tempTodo: Todo | null,
+  setTempTodo: (newTodo: Todo | null) => void,
+}
+
+export const TodosContext = React.createContext<ContextProps>({
+  todos: [],
+  visibleTodos: [],
+  todoAdd: () => new Promise<boolean>(() => {}),
+  clearAllCompleted: () => {},
+  todoDelete: () => new Promise<boolean>(() => {}),
+  isTodosHasCompleted: () => false,
+  isEveryTodoCompleted: () => false,
+  filterBy: Filter.ALL,
+  setFilterBy: () => {},
+  errorMessage: '',
+  areCompletedDeletingNow: false,
+  tempTodo: null,
+  setTempTodo: () => {},
+});
+
+interface ProviderProps {
+  children: React.ReactNode,
+}
+
+export const TodosProvider: React.FC<ProviderProps> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterBy, setFilterBy] = useState(Filter.ALL);
+  const [areCompletedDeletingNow, setAreCompletedDeletingNow]
+    = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const visibleTodos = filterTodos(todos, { filterBy });
+
+  useEffect(() => {
+    getTodos(USER_ID).then(setTodos);
+  }, []);
+
+  const handleErrorOccuring = (errorTitle: string) => {
+    setErrorMessage(errorTitle);
+
+    setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+  };
```

Comment:
might be better to move it to TodoErrors component, because you will use it not only on todo loading
create a timer and clear the timer before the effect runs again
```
useEffect(() => {
    const timeout = setTimeout(closeNotification, 3000);

    return () => clearTimeout(timeout);
  }, [error]);
```